File: frontend/oei-web/src/app/domain/model/profile/professional-profile.ts
export type Availability = 'AVAILABLE' | 'OPEN_TO_OPPORTUNITIES' | 'NOT_AVAILABLE';

export type LanguageLevel = 'A1' | 'A2' | 'B1' | 'B2' | 'C1' | 'C2' | 'NATIVE';

export interface LanguageProficiency {
  readonly language: string;
  readonly level: LanguageLevel;
}

export interface Experience {
  readonly id: string;
  readonly organization: string;
  readonly title: string;
  readonly startDate: string;
  readonly endDate?: string | null;
  readonly current?: boolean;
  readonly description?: string;
  // Demonstration data must always be explicitly flagged so it can never be mistaken
  // for a real member's history — see spec requirement "données de démonstration honnêtes".
  readonly isDemoData?: boolean;
  // The member's own gross annual salary ("salaire brut annuel") for THIS experience.
  // Deliberately private, always — same rule as `CurrentCompensation` below: never part of
  // `PublicProfile.visibleFields`, never rendered on the digital business card/CV/member
  // directory, never sent to any `/api/public/**` endpoint. Saving/updating an experience that
  // carries this field is the trigger for a `compensation_declaration` row feeding the
  // anonymized Professional Neural Network salary-transparency aggregate — never an
  // individually-visible figure. Required together with `salaryCurrency`.
  readonly grossAnnualSalary?: number;
  // ISO 4217 currency code (e.g. "CHF", "EUR") for `grossAnnualSalary` — free text, same
  // convention as `CurrentCompensation.currency`. Required together with `grossAnnualSalary`.
  readonly salaryCurrency?: string;
}

export interface Education {
  readonly id: string;
  readonly institution: string;
  readonly program: string;
  readonly startDate: string;
  readonly endDate?: string | null;
  readonly description?: string;
}

export interface Skill {
  readonly id: string;
  readonly name: string;
  readonly category: string;
  readonly verified?: boolean;
}

// All optional, and every link is a full URL edited as plain text (same "no dedicated
// upload/OAuth pipeline yet" limitation as `photoUrl`). Unlike `currentCompensation` below,
// these ARE meant to be shareable — a member can opt to expose them on their public profile
// via `PublicProfile.visibleFields` (see `public-profile.ts`'s `'socialLinks'` entry).
export interface SocialLinks {
  readonly linkedin?: string;
  readonly github?: string;
  readonly x?: string;
  readonly website?: string;
  readonly youtube?: string;
}

export type CompensationPeriod = 'YEAR' | 'MONTH';

// Deliberately private, always. This field is NEVER part of `PublicProfile.visibleFields`
// (see `PUBLIC_PROFILE_VISIBLE_FIELD_KEYS` in `public-profile.ts` — `currentCompensation` is
// intentionally absent from that list, not just unchecked by default), never rendered on the
// digital business card / CV / member directory, and never sent to any `/api/public/**`
// endpoint. Its purpose is twofold, both aggregate-only uses of the member's own figure, never
// a direct exposure of it:
// - let the member compare their own situation against the anonymized aggregate range computed
//   by `SalaryBenchmarkService` (matched on `expertiseAreas` + `currency` + `period`);
// - feed the Professional Neural Network's anonymized salary transparency feature
//   (`NetworkSalaryInsight`, see `domain/model/network/network-salary-insight.model.ts`), which
//   aggregates declarations by graph node (domain/topic/certification) and, additionally, by
//   `country` below — again only ever surfaced as a computed range once at least
//   `MIN_ANONYMIZED_SAMPLE_SIZE` members contributed to that specific node (+country) pool, never
//   as an individual figure.
// NOTE: the actual write path into `compensation_declaration` is now `Experience.grossAnnualSalary`
// (per-experience, see that field's own doc comment), not this field — this field remains for
// the member's own benchmark comparison only, per the owner's decision to model the trigger at
// the experience level instead of a single profile-wide figure.
export interface CurrentCompensation {
  readonly amount: number;
  // ISO 4217 currency code (e.g. "CHF", "EUR") — free text input, not a hardcoded list, since
  // OEI members span many countries.
  readonly currency: string;
  readonly period: CompensationPeriod;
  // Free-text country label, same format as `NetworkExpert.country` (e.g. "Suisse", "France") —
  // kept consistent with that field rather than an ISO code since the rest of this codebase's
  // country data (network graph demo experts, member directory) is plain localized country
  // names, not codes. Optional: a member can decline to state a country, in which case their
  // declaration only contributes to the country-agnostic (global) aggregate, never to a
  // per-country one.
  readonly country?: string;
}

export interface ProfessionalProfile {
  readonly memberId: string;
  /**
   * Indicates how the profile content was initially obtained.
   * MANUAL: wizard. LINKEDIN_BASIC: LinkedIn identity OAuth.
   * CV_IMPORTED: AI-assisted CV import. LINKEDIN_AND_CV: both.
   * Null for legacy accounts created before the import-first onboarding.
   */
  readonly source?: 'MANUAL' | 'LINKEDIN_BASIC' | 'CV_IMPORTED' | 'LINKEDIN_AND_CV' | null;
  // Set from the onboarding wizard's "Photo" step (`espaceMembre.onboarding.steps.photo`,
  // `onboarding.ts`'s `photoUrl` draft field) or edited later from the profile page itself.
  // No real upload pipeline exists yet — this is a plain URL, exactly like the onboarding
  // step it comes from.
  readonly photoUrl?: string;
  readonly title?: string;
  readonly summary?: string;
  readonly location?: string;
  readonly availability?: Availability;
  readonly expertiseAreas: readonly string[];
  readonly technologies: readonly string[];
  readonly sectors: readonly string[];
  readonly languages: readonly LanguageProficiency[];
  readonly experiences: readonly Experience[];
  readonly educations: readonly Education[];
  readonly skills: readonly Skill[];
  readonly socialLinks?: SocialLinks;
  readonly currentCompensation?: CurrentCompensation;
  readonly completenessScore: number;
}

export function createProfessionalProfile(fields: ProfessionalProfile): ProfessionalProfile {
  return Object.freeze({ ...fields });
}

// Mirrors the fields the completeness score in `ProfessionalProfile.completenessScore`
// is computed from server-side. Kept here (rather than hardcoded in the component) so
// the onboarding wizard and the profile page agree on what "complete" means.
export const PROFILE_COMPLETENESS_FIELDS = [
  'title',
  'summary',
  'location',
  'availability',
  'expertiseAreas',
  'languages',
  'experiences',
  'educations',
  'skills',
] as const;

export function computeCompletenessScore(profile: ProfessionalProfile): number {
  const total = PROFILE_COMPLETENESS_FIELDS.length;
  let filled = 0;
  for (const field of PROFILE_COMPLETENESS_FIELDS) {
    const value = profile[field];
    if (Array.isArray(value) ? value.length > 0 : Boolean(value)) {
      filled += 1;
    }
  }
  return Math.round((filled / total) * 100);
}

File: backend/domain/shared/src/main/java/global/oei/domain/shared/profile/ProfessionalProfile.java
package global.oei.domain.shared.profile;

import java.util.List;
import java.util.Objects;

import global.oei.domain.shared.member.MemberId;
import global.oei.domain.shared.network.CompensationDeclarationCandidate;
import global.oei.domain.shared.network.CompensationPeriod;
import global.oei.domain.shared.network.NetworkSalaryNodeType;

/**
 * A member's professional profile — deliberately a "big object" replaced wholesale by
 * {@code PUT /api/member/v1/profile} rather than exposing per-field CRUD, per ADR 0002
 * ("simplicité du contrat prime sur la granularité").
 *
 * <p>{@link #currentCompensation()} is always private: never part of any public-facing
 * schema/endpoint. See its own Javadoc for how it currently relates (or, more precisely,
 * does not yet automatically relate) to the anonymized {@code SalaryInsight} aggregates.</p>
 *
 * <p>{@link #source()} records how the profile content was initially obtained (manual,
 * LinkedIn, CV import …). A {@code null} source means the member never went through a
 * structured onboarding path — typically a legacy account that predates the import-first
 * onboarding feature. It is NOT a required field: the profile is valid with or without a
 * source value.</p>
 *
 * <p>Note: a {@code ProfessionalProfile} does NOT require a CV. The CV document (see
 * {@link global.oei.domain.shared.cv.Cv}) is an independent, optional construct. A
 * profile may exist without any CV document, including when it was created via
 * LinkedIn or manually.</p>
 */
public record ProfessionalProfile(
        MemberId memberId,
        ProfileSource source,
        String title,
        String summary,
        String location,
        Availability availability,
        List<String> expertiseAreas,
        List<String> technologies,
        List<String> sectors,
        List<LanguageProficiency> languages,
        List<Experience> experiences,
        List<Education> educations,
        List<Skill> skills,
        CurrentCompensation currentCompensation,
        int completenessScore) {

    public ProfessionalProfile {
        Objects.requireNonNull(memberId, "memberId must not be null");
        expertiseAreas = List.copyOf(expertiseAreas == null ? List.of() : expertiseAreas);
        technologies = List.copyOf(technologies == null ? List.of() : technologies);
        sectors = List.copyOf(sectors == null ? List.of() : sectors);
        languages = List.copyOf(languages == null ? List.of() : languages);
        experiences = List.copyOf(experiences == null ? List.of() : experiences);
        educations = List.copyOf(educations == null ? List.of() : educations);
        skills = List.copyOf(skills == null ? List.of() : skills);
        if (completenessScore < 0 || completenessScore > 100) {
            throw new IllegalArgumentException("completenessScore must be between 0 and 100");
        }
    }

    /**
     * @return a new instance with {@link #memberId()} replaced; every other field is
     *         unchanged. Used at the HTTP boundary to overwrite whatever {@code memberId} a
     *         PUT request body carried with the authenticated caller's own identity — a
     *         member can only ever update their own profile.
     */
    public ProfessionalProfile withMemberId(final MemberId memberId) {
        return new ProfessionalProfile(
                memberId,
                source,
                title,
                summary,
                location,
                availability,
                expertiseAreas,
                technologies,
                sectors,
                languages,
                experiences,
                educations,
                skills,
                currentCompensation,
                completenessScore);
    }

    /**
     * @return a new instance with {@link #source()} replaced; every other field is unchanged.
     */
    public ProfessionalProfile withSource(final ProfileSource source) {
        return new ProfessionalProfile(
                memberId,
                source,
                title,
                summary,
                location,
                availability,
                expertiseAreas,
                technologies,
                sectors,
                languages,
                experiences,
                educations,
                skills,
                currentCompensation,
                completenessScore);
    }

    /**
     * Recomputes {@link #completenessScore()} from the fields this profile currently holds,
     * mirroring the frontend's completeness rubric (title/summary/location/availability,
     * >=1 language, >=1 experience, >=1 education, >=3 skills, non-empty expertise areas —
     * each worth an equal share of 100, rounded down).
     *
     * @return a new instance with {@link #completenessScore()} replaced; every other field
     *         is unchanged
     */
    public ProfessionalProfile withRecomputedCompleteness() {
        final int criteriaCount = 6;
        int satisfied = 0;
        if (isNotBlank(title) && isNotBlank(summary) && isNotBlank(location) && availability != null) {
            satisfied++;
        }
        if (!expertiseAreas.isEmpty()) {
            satisfied++;
        }
        if (!languages.isEmpty()) {
            satisfied++;
        }
        if (!experiences.isEmpty()) {
            satisfied++;
        }
        if (!educations.isEmpty()) {
            satisfied++;
        }
        if (skills.size() >= 3) {
            satisfied++;
        }
        final int score = satisfied * 100 / criteriaCount;
        return new ProfessionalProfile(
                memberId,
                source,
                title,
                summary,
                location,
                availability,
                expertiseAreas,
                technologies,
                sectors,
                languages,
                experiences,
                educations,
                skills,
                currentCompensation,
                score);
    }

    private static boolean isNotBlank(final String value) {
        return value != null && !value.isBlank();
    }

    /**
     * Derives the full set of anonymized-pool candidate rows this profile currently feeds,
     * from every {@link Experience} that carries a {@code grossAnnualSalary} (the trigger — see
     * that field's own Javadoc), fanned out across this profile's {@link #expertiseAreas()} as
     * {@link NetworkSalaryNodeType#DOMAIN} graph nodes.
     *
     * <p><strong>Open modeling question</strong> (see MEMBER-SPACE-CURRENT-STATE.md §6): a
     * member's freeform {@code expertiseAreas} entries are not validated against the canonical
     * {@code NetworkDomain} catalog (fixed slugs such as {@code "ia"}/{@code "cloud"}); a
     * declaration whose {@code nodeId} does not match a real domain slug is stored but never
     * surfaces in any {@code SalaryInsight} aggregate. A dedicated expertise-area-to-domain
     * mapping (or a domain picker constrained to the catalog) would close this gap; out of
     * scope here since {@code Experience} itself carries no domain/topic linkage today.</p>
     *
     * @param country free-text country label (same format as {@code CurrentCompensation.country}
     *                 / {@code NetworkExpert.country}), or {@code null} to leave every derived
     *                 declaration country-agnostic
     * @return one candidate per (salaried experience × expertise area) pair; empty when no
     *         experience carries a salary or the profile lists no expertise area to attach to
     */
    public List<CompensationDeclarationCandidate> deriveCompensationDeclarations(final String country) {
        if (expertiseAreas.isEmpty()) {
            return List.of();
        }
        return experiences.stream()
                .filter(Experience::hasGrossAnnualSalary)
                .flatMap(experience -> expertiseAreas.stream()
                        .map(domainId -> new CompensationDeclarationCandidate(
                                NetworkSalaryNodeType.DOMAIN,
                                domainId,
                                country,
                                experience.grossAnnualSalary(),
                                experience.salaryCurrency(),
                                CompensationPeriod.YEAR)))
                .toList();
    }
}
